# Equipo: 10
Grupo: 002 Miércoles


### Integrantes
- 1864878 Cerda Serrato Alan René
- 1867341 Cruz Lerma Jorge Antonio
- 1941475 Galindo Vázquez Jesús Alfonso
- 1630606 Valencia González Jesús Eduardo


# Netflix Movies and TV Shows
URL: https://www.kaggle.com/shivamb/netflix-shows

### a) Importar librerías y base de datos.

In [1]:
#Importar la librería necesaria
import pandas as pd
import numpy as np

#Importar la base de datos
net = pd.read_csv("/content/netflix_titles.csv")

net #Comprobamos que en efecto se importó la base de datos

show_id  ...                                        description
0         s1  ...  In a future where the elite inhabit an island ...
1         s2  ...  After a devastating earthquake hits Mexico Cit...
2         s3  ...  When an army recruit is found dead, his fellow...
3         s4  ...  In a postapocalyptic world, rag-doll robots hi...
4         s5  ...  A brilliant group of students become card-coun...
...      ...  ...                                                ...
7782   s7783  ...  When Lebanon's Civil War deprives Zozo of his ...
7783   s7784  ...  A scrappy but poor boy worms his way into a ty...
7784   s7785  ...  In this documentary, South African rapper Nast...
7785   s7786  ...  Dessert wizard Adriano Zumbo looks for the nex...
7786   s7787  ...  This documentary delves into the mystique behi...

[7787 rows x 12 columns]

Ya se importaron las librerías y la base de datos.

### b) Visualizar los datos.

- Conocer el tamaño de la base:

In [2]:
#Conteo de filas y columnas
net.shape

(7787, 12)

Sabemos que la base de datos tiene 7787 filas y 12 columnas.

- Primeras y últimas 3 filas de la base de datos:

In [3]:
#Primeras 3 filas
net.head(3)

show_id  ...                                        description
0      s1  ...  In a future where the elite inhabit an island ...
1      s2  ...  After a devastating earthquake hits Mexico Cit...
2      s3  ...  When an army recruit is found dead, his fellow...

[3 rows x 12 columns]

In [4]:
#Últimas 3 filas
net.tail(3)

show_id  ...                                        description
7784   s7785  ...  In this documentary, South African rapper Nast...
7785   s7786  ...  Dessert wizard Adriano Zumbo looks for the nex...
7786   s7787  ...  This documentary delves into the mystique behi...

[3 rows x 12 columns]

### c) Descripción básica de columnas

- Nombres de las columnas:

In [5]:
net.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

- Tipo de datos en las columnas:

In [6]:
net.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Como vemos, a excepción de la columna "release_year" que tiene datos enteros, las columnas son del tipo objeto.

- Datos nulos

In [7]:
net.isnull().any()

show_id         False
type            False
title           False
director         True
cast             True
country          True
date_added       True
release_year    False
rating           True
duration        False
listed_in       False
description     False
dtype: bool

Como vemos, hay varias columnas con datos nulos (True). Ahora veamos cuántos datos nulos hay en cada columna:

In [8]:
net.isnull().sum()

show_id            0
type               0
title              0
director        2389
cast             718
country          507
date_added        10
release_year       0
rating             7
duration           0
listed_in          0
description        0
dtype: int64

Podemos apreciar que hay muchos datos nulos. Sobre todo en la columna "director", con 2389 datos nulos.

### d) y e) Clasificación de los datos y renombrar columnas

- Eliminar las columnas que no son necesarias

In [9]:
#Ahora que conocemos los nombres, eliminamos las columnas no deseadas
net.drop('show_id', axis = 1, inplace = True)
net.drop('director', axis = 1, inplace = True)
net.drop('cast', axis = 1, inplace = True)
net.drop('country', axis = 1, inplace = True)
net.drop('date_added', axis = 1, inplace = True)
net.drop('release_year', axis = 1, inplace = True)

net.columns  #Comprobamos que en efecto se eliminaron las columnas no deseadas

Index(['type', 'title', 'rating', 'duration', 'listed_in', 'description'], dtype='object')

Ya que tenemos solo las columnas que deseamos, procedemos a renombrarlas para que se entienda mejor qué muestra cada una:

In [10]:
net.rename(columns = {'type':'Tipo'}, inplace = True)
net.rename(columns = {'title':'Nombre'}, inplace = True)
net.rename(columns = {'rating':'Clasificación'}, inplace = True)
net.rename(columns = {'duration':'Duración'}, inplace = True)
net.rename(columns = {'listed_in':'Género'}, inplace = True)
net.rename(columns = {'description':'Información'}, inplace = True)

net.columns   #Comprobamos que se hayan cambiado los nombres

Index(['Tipo', 'Nombre', 'Clasificación', 'Duración', 'Género', 'Información'], dtype='object')

Ahora los clasificamos por tipo (si es serie o película):

In [11]:
net.Tipo.unique()

array(['TV Show', 'Movie'], dtype=object)

Solo hay 2 "Tipos": Series y películas (TV Show, Movie)

In [12]:
Series = net[net['Tipo'] == 'TV Show']
Pelis = net[net['Tipo'] == 'Movie']

Ya están separados los arreglos. Ahora comprobamos que se separaron correctamente.

Comenzamos con las series:

In [13]:
Series.Tipo.unique()

array(['TV Show'], dtype=object)

Como vemos, en el arreglo de series, solo se encuentran las filas "TV Show". Este arreglo parece correcto.

Ahora seguimos con el arreglo de las películas:

In [14]:
Pelis.Tipo.unique()

array(['Movie'], dtype=object)

También está correcto. Solo están las películas.

Y como ya vimos que la columna "type" que fue renombrada a "Tipo" no tiene datos nulos, podemos estar seguros de que todas las películas y series fueron clasificados como corresponde.

### f) Guardar las agrupaciones en archivos

In [15]:
Pelis.to_csv('Películas de Netflix.csv', index = False)
Series.to_csv('Series de Netflix.csv', index = False)

Ahora, solo queda asegurarse de que los archivos tienen los datos correctos.

Primero el archivo de películas:

In [16]:
P = pd.read_csv('Películas de Netflix.csv')
P

Tipo  ...                                        Información
0     Movie  ...  After a devastating earthquake hits Mexico Cit...
1     Movie  ...  When an army recruit is found dead, his fellow...
2     Movie  ...  In a postapocalyptic world, rag-doll robots hi...
3     Movie  ...  A brilliant group of students become card-coun...
4     Movie  ...  After an awful accident, a couple admitted to ...
...     ...  ...                                                ...
5372  Movie  ...  Dragged from civilian life, a former superhero...
5373  Movie  ...  When Lebanon's Civil War deprives Zozo of his ...
5374  Movie  ...  A scrappy but poor boy worms his way into a ty...
5375  Movie  ...  In this documentary, South African rapper Nast...
5376  Movie  ...  This documentary delves into the mystique behi...

[5377 rows x 6 columns]

Ahora el de series:

In [17]:
S = pd.read_csv('Series de Netflix.csv')
S

Tipo  ...                                        Información
0     TV Show  ...  In a future where the elite inhabit an island ...
1     TV Show  ...  A genetics professor experiments with a treatm...
2     TV Show  ...  In this dark alt-history thriller, a naïve law...
3     TV Show  ...  Archival video and new interviews examine Mexi...
4     TV Show  ...  As a psychology professor faces Alzheimer's, h...
...       ...  ...                                                ...
2405  TV Show  ...  Strong-willed, middle-class Kashaf and carefre...
2406  TV Show  ...  A quest for freedom and legendary treasure beg...
2407  TV Show  ...  While living alone in a spooky town, a young g...
2408  TV Show  ...  An assortment of talent takes the stage for a ...
2409  TV Show  ...  Dessert wizard Adriano Zumbo looks for the nex...

[2410 rows x 6 columns]

Como podemos ver, ambos archivos tienen los datos correctos.

# Preguntas de interés

**¿Para que me puede servir estas agrupaciones para la práctica siguiente?**

Con estas agrupaciones será más sencillo encontrar películas y series con ciertas especificaciones, por ejemplo, series con 3 temporadas, o películas que duren menos de 2 horas.

**Importancia del analisis previo de los datos antes de iniciar las modificaciones.**

Al analizar los datos antes de modificar, nos podemos dar cuenta de si la base "está como queremos". Así podemos evitar hacer modificaciones innecesarias, y nos podemos dar cuenta qué partes en específico necesitan modificarse.

# Puntos extra

## Investigar los comandos df.dropna() y df.astype()

#### **df.dropna():** Permite eliminar filas o columnas con datos nulos.

Por ejemplo, eliminemos las filas con datos nulos en el archivo de las películas:

In [18]:
#Primero veamos si hay datos nulos
P.isnull().any()

Tipo             False
Nombre           False
Clasificación     True
Duración         False
Género           False
Información      False
dtype: bool

In [19]:
#Como vemos, hay 7 datos nulos en la columna clasificación. Ahora veamos cuáles son las filas con los datos nulos
P[P.isnull().any(True)]

Tipo  ...                                        Información
51    Movie  ...  Oprah Winfrey sits down with director Ava DuVe...
2611  Movie  ...  Louis C.K. muses on religion, eternal love, gi...
2612  Movie  ...  Emmy-winning comedy writer Louis C.K. brings h...
2613  Movie  ...  The comic puts his trademark hilarious/thought...
3014  Movie  ...  Amid the chaos and horror of World War II, a c...

[5 rows x 6 columns]

In [20]:
#Ahora eliminemos las filas con datos nulos
a = P.dropna() #Asigno variable nueva para guardar las modificaciones sin sobreescribir los datos de P
#Comprobemos si los datos de 2611 al 2613 ya no están
a[2609:2614]
#Como vemos, las filas con datos nulos se eliminaron

Tipo  ...                                        Información
2610  Movie  ...  Grammy-winning singer Loudon Wainwright III re...
2614  Movie  ...  When professional ambitions clash with persona...
2615  Movie  ...  A sexually inexperienced rickshaw driver strug...
2616  Movie  ...  After enrolling in a college poetry course, a ...
2617  Movie  ...  Sparks fly and romance blooms when Shelby, a v...

[5 rows x 6 columns]

Ahora eliminemos las columnas con datos nulos en el archivo de series:

In [21]:
#Veamos si hay datos nulos
S.isnull().any()

Tipo             False
Nombre           False
Clasificación     True
Duración         False
Género           False
Información      False
dtype: bool

In [22]:
#Ahora eliminemos las columnas con datos nulos y mostremos la base
S.dropna(axis = 1)
#Como vemos, la columna Clasificación ya no está por tener datos nulos.

Tipo  ...                                        Información
0     TV Show  ...  In a future where the elite inhabit an island ...
1     TV Show  ...  A genetics professor experiments with a treatm...
2     TV Show  ...  In this dark alt-history thriller, a naïve law...
3     TV Show  ...  Archival video and new interviews examine Mexi...
4     TV Show  ...  As a psychology professor faces Alzheimer's, h...
...       ...  ...                                                ...
2405  TV Show  ...  Strong-willed, middle-class Kashaf and carefre...
2406  TV Show  ...  A quest for freedom and legendary treasure beg...
2407  TV Show  ...  While living alone in a spooky town, a young g...
2408  TV Show  ...  An assortment of talent takes the stage for a ...
2409  TV Show  ...  Dessert wizard Adriano Zumbo looks for the nex...

[2410 rows x 5 columns]

#### **df.astype():** Cambia el tipo de datos de los objetos a un tipo de datos especificado.

Por ejemplo, cambiemos el tipo de datos de la base principal al tipo bool:

In [23]:
#Tipo de datos original
net.dtypes

Tipo             object
Nombre           object
Clasificación    object
Duración         object
Género           object
Información      object
dtype: object

In [24]:
#Cambio del tipo de datos de toda la base
c = net.astype(bool)
c.dtypes

Tipo             bool
Nombre           bool
Clasificación    bool
Duración         bool
Género           bool
Información      bool
dtype: object

In [25]:
#Si queremos que solo una columna cambie de tipo, entonces:
d = net.Duración.astype(bool)
d.head()

0    True
1    True
2    True
3    True
4    True
Name: Duración, dtype: bool

## Tarea de preprocesamiento no mencionada en la práctica

Separar los datos de alguna columna y generar nuevas columnas con los datos separados.

In [47]:
#Los géneros están separados por comas y quiero crear nuevas columnas
x = net["Género"].str.split(",", expand = True)   #https://www.analyticslane.com/2020/10/19/separar-texto-en-columnas-con-pandas-en-python/
x

0                      1                     2
0     International TV Shows              TV Dramas   TV Sci-Fi & Fantasy
1                     Dramas   International Movies                  None
2              Horror Movies   International Movies                  None
3         Action & Adventure     Independent Movies      Sci-Fi & Fantasy
4                     Dramas                   None                  None
...                      ...                    ...                   ...
7782                  Dramas   International Movies                  None
7783                  Dramas   International Movies      Music & Musicals
7784           Documentaries   International Movies      Music & Musicals
7785  International TV Shows             Reality TV                  None
7786           Documentaries       Music & Musicals                  None

[7787 rows x 3 columns]

In [54]:
#Ahora quiero renombrar esas columnas
x.columns = ['Género 1', 'Género 2', 'Género 3']
x

Género 1               Género 2              Género 3
0     International TV Shows              TV Dramas   TV Sci-Fi & Fantasy
1                     Dramas   International Movies                  None
2              Horror Movies   International Movies                  None
3         Action & Adventure     Independent Movies      Sci-Fi & Fantasy
4                     Dramas                   None                  None
...                      ...                    ...                   ...
7782                  Dramas   International Movies                  None
7783                  Dramas   International Movies      Music & Musicals
7784           Documentaries   International Movies      Music & Musicals
7785  International TV Shows             Reality TV                  None
7786           Documentaries       Music & Musicals                  None

[7787 rows x 3 columns]

In [58]:
#Ahora quiero que estas nuevas columnas estén en la base de datos principal
net2 = pd.concat([net, x], axis = 1)
net2

Tipo  ...              Género 3
0     TV Show  ...   TV Sci-Fi & Fantasy
1       Movie  ...                  None
2       Movie  ...                  None
3       Movie  ...      Sci-Fi & Fantasy
4       Movie  ...                  None
...       ...  ...                   ...
7782    Movie  ...                  None
7783    Movie  ...      Music & Musicals
7784    Movie  ...      Music & Musicals
7785  TV Show  ...                  None
7786    Movie  ...                  None

[7787 rows x 9 columns]

In [62]:
#Se ve feo que diga "None", así que cambiaré esos nombres por un guión
net2.fillna(value = "-")

Tipo  ...              Género 3
0     TV Show  ...   TV Sci-Fi & Fantasy
1       Movie  ...                     -
2       Movie  ...                     -
3       Movie  ...      Sci-Fi & Fantasy
4       Movie  ...                     -
...       ...  ...                   ...
7782    Movie  ...                     -
7783    Movie  ...      Music & Musicals
7784    Movie  ...      Music & Musicals
7785  TV Show  ...                     -
7786    Movie  ...                     -

[7787 rows x 9 columns]